Add SPI for connecting to webpushd

szewai opened the pull request:
#### 9f0d6c1f3935637aa86546cc0643676f0b34af41
<pre>
Add SPI for connecting to webpushd
<a href="https://bugs.webkit.org/show_bug.cgi?id=279142">https://bugs.webkit.org/show_bug.cgi?id=279142</a>
<a href="https://rdar.apple.com/135297138">rdar://135297138</a>

Reviewed by Ben Nham.

This will make it easier to test webpushd functionalities.

Tests: WebPushD.WKWebPushDaemonConnectionRequestPushPermission
       WebPushD.WKWebPushDaemonConnectionPushSubscription

* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.cpp: Added.
(API::WebPushDaemonConnection::WebPushDaemonConnection):
(API::WebPushDaemonConnection::getPushPermissionState):
(API::WebPushDaemonConnection::requestPushPermission):
(API::WebPushDaemonConnection::setAppBadge):
(API::WebPushDaemonConnection::subscribeToPushService):
(API::WebPushDaemonConnection::unsubscribeFromPushService):
(API::WebPushDaemonConnection::getPushSubscription):
(API::WebPushDaemonConnection::getNextPendingPushMessage):
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.h: Added.
* Source/WebKit/UIProcess/API/APIWebPushMessage.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/APIWebPushSubscriptionData.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm: Added.
(-[_WKWebPushDaemonConnectionConfiguration init]):
(-[_WKWebPushDaemonConnection initWithConfiguration:]):
(toWKPermissionsState):
(-[_WKWebPushDaemonConnection getPushPermissionStateForOrigin:completionHandler:]):
(-[_WKWebPushDaemonConnection requestPushPermissionForOrigin:completionHandler:]):
(-[_WKWebPushDaemonConnection setAppBadge:origin:]):
(-[_WKWebPushDaemonConnection subscribeToPushServiceForScope:applicationServerKey:completionHandler:]):
(-[_WKWebPushDaemonConnection unsubscribeFromPushServiceForScope:completionHandler:]):
(-[_WKWebPushDaemonConnection getSubscriptionForScope:completionHandler:]):
(-[_WKWebPushDaemonConnection getNextPendingPushMessage:]):
(-[_WKWebPushDaemonConnection _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnectionInternal.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushMessage.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushMessage.mm: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
(-[_WKWebPushMessage dealloc]):
(-[_WKWebPushMessage data]):
(-[_WKWebPushMessage scope]):
(-[_WKWebPushMessage partition]):
(-[_WKWebPushMessage _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushMessageInternal.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionData.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionData.mm: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
(-[_WKWebPushSubscriptionData dealloc]):
(-[_WKWebPushSubscriptionData endpoint]):
(-[_WKWebPushSubscriptionData applicationServerKey]):
(-[_WKWebPushSubscriptionData authenticationSecret]):
(-[_WKWebPushSubscriptionData ecdhPublicKey]):
(-[_WKWebPushSubscriptionData _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionDataInternal.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionRequestPushPermission)):
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionPushSubscription)):

Canonical link: <a href="https://commits.webkit.org/283218@main">https://commits.webkit.org/283218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/374bc8620d3bbcae47ebc829d7530d062034b6a6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65543 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69569 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16152 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52715 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16432 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52626 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11204 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68610 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41482 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56727 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33254 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38164 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14105 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15028 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14447 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71275 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9497 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13894 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59979 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9529 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56792 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60217 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14461 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7846 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1491 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40724 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41800 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42983 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41544 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.h:
```diff
@@ -0,0 +1,66 @@
+    void getNextPendingPushMessage(CompletionHandler<void(const std::optional<WebKit::WebPushMessage>&)>&&);
+
+private:
+    std::unique_ptr<WebKit::WebPushD::Connection> m_connection;
```
I think this could likely be a UniqueRef.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h:
```diff
@@ -0,0 +1,66 @@
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithConfiguration:(_WKWebPushDaemonConnectionConfiguration *)configuration;
+
+- (void)getPushPermissionStateForScope:(NSURL *)scopeURL completionHandler:(void (^)(_WKWebPushPermissionState))completionHandler;
```
I think this should be named `getPushPermissionStateForOrigin:` since the granularity of permission is an origin and I don't think we have any plans on changing that.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h:
```diff
@@ -0,0 +1,66 @@
+- (instancetype)initWithConfiguration:(_WKWebPushDaemonConnectionConfiguration *)configuration;
+
+- (void)getPushPermissionStateForScope:(NSURL *)scopeURL completionHandler:(void (^)(_WKWebPushPermissionState))completionHandler;
+- (void)requestPushPermissionForScope:(NSURL *)scopeURL completionHandler:(void (^)(BOOL))completionHandler;
```
Similarly I think this should be named `requestPushPermissionForOrigin:`.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h:
```diff
@@ -0,0 +1,66 @@
+
+- (void)getPushPermissionStateForScope:(NSURL *)scopeURL completionHandler:(void (^)(_WKWebPushPermissionState))completionHandler;
+- (void)requestPushPermissionForScope:(NSURL *)scopeURL completionHandler:(void (^)(BOOL))completionHandler;
+- (void)setAppBadge:(NSUInteger *)badge scopeURL:(NSURL *)scopeURL;
```
Probably the `scopeURL` here should also be renamed `origin`.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
```diff
@@ -0,0 +1,157 @@
+        if (result)
+            return completionHandlerCopy(wrapper(API::WebPushSubscriptionData::create(WTFMove(result.value()))).get(), nil);
+
+        auto error = [NSError errorWithDomain:@"WKErrorDomain" code:WKErrorUnknown userInfo:@{ NSLocalizedDescriptionKey:result.error().message }];
```
I think this is okay for now but maybe in the future we should provide more info in the error since these are literally DOMExceptions. For instance, we can define an error code `WKErrorCodeDOMException` and a userInfo keys `WKErrorCodeDOMExceptionNameKey` and `WKErrorCodeDOMExceptionMessageKey` which contains a string representation of the DOMException name and message respectively. Can be deferred to a future task.

szewai replied:
Sure; I left a FIXME comment here.

bnham reviewed (commented):
I left a few nits but this mostly LGTM.

bnham reviewed (approved):
LGTM

szewai committed:
Add SPI for connecting to webpushd
https://bugs.webkit.org/show_bug.cgi?id=279142
rdar://135297138

Reviewed by Ben Nham.

This will make it easier to test webpushd functionalities.

Tests: WebPushD.WKWebPushDaemonConnectionRequestPushPermission
       WebPushD.WKWebPushDaemonConnectionPushSubscription

* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.cpp: Added.
(API::WebPushDaemonConnection::WebPushDaemonConnection):
(API::WebPushDaemonConnection::getPushPermissionState):
(API::WebPushDaemonConnection::requestPushPermission):
(API::WebPushDaemonConnection::setAppBadge):
(API::WebPushDaemonConnection::subscribeToPushService):
(API::WebPushDaemonConnection::unsubscribeFromPushService):
(API::WebPushDaemonConnection::getPushSubscription):
(API::WebPushDaemonConnection::getNextPendingPushMessage):
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.h: Added.
* Source/WebKit/UIProcess/API/APIWebPushMessage.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/APIWebPushSubscriptionData.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm: Added.
(-[_WKWebPushDaemonConnectionConfiguration init]):
(-[_WKWebPushDaemonConnection initWithConfiguration:]):
(toWKPermissionsState):
(-[_WKWebPushDaemonConnection getPushPermissionStateForOrigin:completionHandler:]):
(-[_WKWebPushDaemonConnection requestPushPermissionForOrigin:completionHandler:]):
(-[_WKWebPushDaemonConnection setAppBadge:origin:]):
(-[_WKWebPushDaemonConnection subscribeToPushServiceForScope:applicationServerKey:completionHandler:]):
(-[_WKWebPushDaemonConnection unsubscribeFromPushServiceForScope:completionHandler:]):
(-[_WKWebPushDaemonConnection getSubscriptionForScope:completionHandler:]):
(-[_WKWebPushDaemonConnection getNextPendingPushMessage:]):
(-[_WKWebPushDaemonConnection _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnectionInternal.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushMessage.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushMessage.mm: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
(-[_WKWebPushMessage dealloc]):
(-[_WKWebPushMessage data]):
(-[_WKWebPushMessage scope]):
(-[_WKWebPushMessage partition]):
(-[_WKWebPushMessage _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushMessageInternal.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionData.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionData.mm: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
(-[_WKWebPushSubscriptionData dealloc]):
(-[_WKWebPushSubscriptionData endpoint]):
(-[_WKWebPushSubscriptionData applicationServerKey]):
(-[_WKWebPushSubscriptionData authenticationSecret]):
(-[_WKWebPushSubscriptionData ecdhPublicKey]):
(-[_WKWebPushSubscriptionData _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionDataInternal.h: Copied from Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h.
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionRequestPushPermission)):
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionPushSubscription)):

Canonical link: https://commits.webkit.org/283218@main
